kanidm: BindMount certificate paths

Flakebi opened the pull request:
###### Description of changes

Bind mount the base dirs of the tls key and chain into the service.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

erictapen commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -224,6 +225,7 @@ in
         Group = "kanidm";
 
         AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+        BindReadOnlyPaths = defaultServiceConfig.BindReadOnlyPaths ++ certPaths;
```
Could we use `lib.mkMerge` for the attrset instead of `//`? That would make this construct obsolete.

erictapen reviewed (commented):
This seems to break the test currently.

Flakebi commented:
Changed to use `mkMerge` ([diff](https://github.com/NixOS/nixpkgs/compare/fe86b196691a54de24af78d5c234fd4a19948a93..ff7a7caf4ac723c0b936ce9175ef5952134c6b5c)).

> This seems to break the test currently.

That is weird, the test passes for me locally on all commits (I’m using `nix-build -A kanidm.tests`).

erictapen commented:
@Flakebi Are you sure that triggers the NixOS integration test? I get this derivation on `x86_64-linux`:

```
$ nix path-info --derivation .#nixosTests.kanidm
/nix/store/6s2zld2iafz42hhwkja4dfh4fgxiwxdk-vm-test-run-kanidm.drv
```

mweinelt commented:
Hi, can we revisit this? Also alpha11 was tagged two days ago.

mweinelt commented:
Please remove the merge commit before we merge this.

Flakebi commented:
I still don’t know why the test passes on my PC but fails on ofborg.
The added commit merges bind paths, so they don’t contain any duplicates or nested paths.
As a consequence, the test should run as before (as the cert paths for the test are inside /nix/store, which is already part of BindReadOnlyPaths).

Flakebi commented:
Ofborg now passes the test as well.

> Please remove the merge commit before we merge this.

I rebased and removed the merge commit.

> > `nix-build -A kanidm.tests`
>
> Are you sure that triggers the NixOS integration test?

Yes it does :)

Flakebi committed:
kanidm: BindMount certificate paths

Bind mount the base dirs of the tls key and chain into the service.

Make sure to bind every directory just once. The test failed on ofborg
when /nix/store and the certificate path in /nix/store/<some path> were
bound.

Flakebi commented:
Thanks for the heads-up! It’s fixed now.

I didn’t see the warnings locally when using `nom-build`, `nix-build` did show them though.

erictapen reviewed (approved):
Excellent, thanks a lot!

dotlambda commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -216,22 +228,28 @@ in
+          AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+          CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" ];
+          # This would otherwise override the CAP_NET_BIND_SERVICE capability.
+          PrivateUsers = lib.mkForce false;
```
Can we at least use `mkOverride 51` instead? This makes it unnecessarily hard to override.
